New example - custom axis tickformatter

lbrito1 opened the pull request:
issues: fixes #3193

By defining a custom TickFormatter model, we are able to remove the weekend/holiday gaps in the candlestick financial chart. Dates are formatted into an array of Strings, e.g. ['Sep 01', 'Sep 02', ...], which is passed to the custom ticker as a property and then rendered on the axis.

The downside to this approach is that we lose the dynamic nature of Datetimei. This can be overcome by making DateGapTickFormatter receive more than one list of Strings, each one representing a different Date format, e.g. List 1 = ['Sep 01', 'Sep 02', ...]; List 2 = ['Sep', 'Sep', ... 'Oct', ...] and so on. This remains as a suggestion only and is not implemented by this commit.

bryevdv commented:
@lbrito1 quote note, our automated image diff testing machinery assumes the output filename matches the python filename, e.g. `foo.py -> foo.html` I think if you make that change the automated tests will pass.

lbrito1 commented:
@bryevdv Thanks, fixed that.

mattpap commented on examples/plotting/file/custom_datetime_axis.py:
```diff
@@ -0,0 +1,58 @@
+		"""
+
+df = pd.DataFrame(MSFT)[:50]
+df = df.set_index([range(0,len(df))])
```
You should wrap `range(...)` in `list()` for python 3, or maybe use `six`.

bryevdv replied:
Yup this is causing the one remaining test failure: https://travis-ci.org/bokeh/bokeh/jobs/96149634#L1032

mattpap commented on examples/plotting/file/custom_datetime_axis.py:
```diff
@@ -0,0 +1,58 @@
+
+df = pd.DataFrame(MSFT)[:50]
+df = df.set_index([range(0,len(df))])
+date_labels = map(lambda x: x.strftime('%b %d'), pd.to_datetime(df["date"]))
```
List comprehension would be much more readable here.

bryevdv commented:
One other suggestion I think some of the description here should go in either a comment at the top of the example or in comments spread throughout.

lbrito1 commented:
Thanks for the suggestions @bryevdv . I realized the MSFT df used in the example actually has integer indices by default, so `df.set_index([range(0,len(df))])` is not necessary. I removed that line and added a note that in case df has an index of a type other than integer, say datetime, it should be set as an integer sequence.

I also changed the lambda map to list comprehension and added some comments throughout the example.

bryevdv commented:
LGTM @bokeh/core any final thoughts? I will merge tomorrow if not.

mattpap commented on examples/plotting/file/custom_datetime_axis.py:
```diff
@@ -0,0 +1,65 @@
+# Strings (e.g. ['Sep 01', 'Sep 02', ...]) passed to the date_labels property. 
+class DateGapTickFormatter(TickFormatter):
+  date_labels = List(item_type=String)
+  __implementation__ = """
```
1. Indentation is off, should be 4 spaces.
2. For readability, one empty line before `__implementation__` should be added.
3. `item_type` is unnecessary, `List(String)` is the proper style (which should resemble scala's `List[String]` syntax).

mattpap commented on examples/plotting/file/custom_datetime_axis.py:
```diff
@@ -0,0 +1,65 @@
+    labels = ( 
+      for tick, i in ticks
+        date_labels[tick] ? ""
+      )
```
As this is a one-liner, it would be better to use `(date_labels[tick] ? "" for tick in ticks)` syntax. Note also `i` is not used, so can be removed.

mattpap commented:
I pointed out some minor stylistic complaints. Otherwise LGTM.

lbrito1 commented:
@mattpap Nice, I wasn't aware of Coffeescript's one-line loop syntax. I updated the branch with your suggestions.

bryevdv commented:
Thanks for the great PR @lbrito1 !
